File: behavioral/go/strategy/strategy_test.go
package strategy_test

import (
	"testing"

	"github.com/AnhellO/td-design-patterns/behavioral/go/strategy"
	"github.com/brianvoe/gofakeit/v6"
)

func init() {
	gofakeit.Seed(1000)
}

func TestDefaults(t *testing.T) {
	tcases := []struct {
		items    []uint8
		expected string
	}{
		{
			make([]uint8, 100),
			"Ordenando con bubble sort!",
		},
		{
			make([]uint8, 10000),
			"Ordenando con merge sort!",
		},
		{
			make([]uint8, 100000),
			"Ordenando con quick sort!",
		},
	}

	for _, tc := range tcases {
		gofakeit.Slice(&tc.items)
		context := strategy.NewSortContext(tc.items, nil)
		got := context.SortContext()

		if got != tc.expected {
			t.Errorf("Expected %s, got %s", tc.expected, got)
		}
	}
}

func TestSpecificStrategy(t *testing.T) {
	tcases := []struct {
		st       strategy.SortStrategy
		expected string
	}{
		{
			&strategy.BubbleSortConcreteStrategy{},
			"Ordenando con bubble sort!",
		},
		{
			&strategy.MergeSortConcreteStrategy{},
			"Ordenando con merge sort!",
		},
		{
			&strategy.QuickSortConcreteStrategy{},
			"Ordenando con quick sort!",
		},
	}

	for _, tc := range tcases {
		context := strategy.NewSortContext([]uint8{1, 9, 8, 3, 55}, tc.st)
		got := context.SortContext()

		if got != tc.expected {
			t.Errorf("Expected %s, got %s", tc.expected, got)
		}
	}
}

func TestUpdatingStrategy(t *testing.T) {
	tcases := []struct {
		concreteStrategy1 strategy.SortStrategy
		concreteStrategy2 strategy.SortStrategy
		expected1         string
		expected2         string
	}{
		{
			&strategy.BubbleSortConcreteStrategy{},
			&strategy.MergeSortConcreteStrategy{},
			"Ordenando con bubble sort!",
			"Ordenando con merge sort!",
		},
	}

	for _, tc := range tcases {
		context := strategy.NewSortContext([]uint8{1, 9, 8, 3, 55}, tc.concreteStrategy1)
		got := context.SortContext()

		if got != tc.expected1 {
			t.Errorf("Expected %s, got %s", tc.expected1, got)
		}

		context.SetStrategy(tc.concreteStrategy2)
		got = context.SortContext()

		if got != tc.expected2 {
			t.Errorf("Expected %s, got %s", tc.expected2, got)
		}
	}
}

File: behavioral/php/strategy/StrategyTest.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;

/**
 * Execute test class from the project directory with the following command:
 * './vendor/bin/phpunit --bootstrap behavioral/php/strategy/Strategy.php behavioral/php/strategy/StrategyTest.php'
 */
class StrategyTest extends TestCase
{
    /**
     * @dataProvider defaultsProvider
     */
    public function testDefaultStrategies(array $items, string $expected): void
    {
        $context = new SortContext($items);
        $this->assertEquals($context->sortContext(), $expected);
    }

    /**
     * @dataProvider specificsProvider
     */
    public function testConcreteStrategy(SortStrategy $concreteStrategy, string $expected): void
    {
        $context = new SortContext([1, 9, 8, 3, 55], $concreteStrategy);
        $this->assertEquals($context->sortContext(), $expected);
    }


    public function testUpdatingStrategy(): void
    {
        $strategyOne = new BubbleSortConcreteStrategy();
        $context = new SortContext([1, 9, 8, 3, 55], $strategyOne);
        $this->assertEquals($context->sortContext(), "Sorting with bubble sort!");
        
        $strategyTwo = new MergeSortConcreteStrategy();
        $context->setStrategy($strategyTwo);
        $this->assertEquals($context->sortContext(), "Sorting with merge sort!");
    }


    public function defaultsProvider(): array
    {
        return [
            [
                range(1, 100),
                "Sorting with bubble sort!"
            ],
            [
                range(1, 10000),
                "Sorting with merge sort!"
            ],
            [
                range(1, 100000),
                "Sorting with quick sort!"
            ],
        ];
    }


    public function specificsProvider(): array
    {
        return [
            [
                new BubbleSortConcreteStrategy([1, 9, 8, 3, 55]),
                "Sorting with bubble sort!"
            ],
            [
                new MergeSortConcreteStrategy([1, 9, 8, 3, 55]),
                "Sorting with merge sort!"
            ],
            [
                new QuickSortConcreteStrategy([1, 9, 8, 3, 55]),
                "Sorting with quick sort!"
            ]
        ];
    }
}